README
This is GNU Mcron, a tool to run jobs at scheduled times.  It is a complete
replacement for Vixie cron.  Besides supporting the traditional Vixie syntax
for its configuration files, GNU Mcron offers the possibility to define jobs
using the Scheme language.

See the INSTALL file for generic information about how to configure and
install GNU Mcron.  If this file is not present, see HACKING for
preliminary build instructions.

----------------------------------------------------------------------
IMPORTANT NOTICES

Do not (yet) install this software on a machine which relies for its
functioning on its current set of crontabs.

To not replace the cron daemon on a system, the package must be installed
with the --disable-multi-user configure option.

Before installing this package for the first time, it is necessary to terminate
any running cron daemons on your system. If your old cron is not Vixie or
accurately Vixie compatible (files in /var/cron/tabs*, /var/cron/allow,
/var/cron/deny, /etc/crontab, /var/run/cron.pid) then you will need to clear out
all old crontabs and make new ones afresh - or else look very carefully at the
options you pass to the package configure script, as follows.

It is often the case that GNU/Linux distributions and other Unices hacked the
cron daemon to use different directories to those above. You can use configure
options --spool-dir, --socket-file, --allow-file, --deny-file, --pid-file and
--tmp-dir to make mcron behave similarly. Note that, with the exception of
tmp-dir, none of these files or directories should be accessible by ordinary
(non-root) users.

If your old cron is Vixie, or very similar, mcron should fall right into place
where your old cron was (the binaries cron and crontab will be replaced, but if
your existing system has a binary called crond, you should make this a link
to mcron), and you should be able to continue to use your existing crontabs
without noticing any changes.

If you don't want to clobber your existing cron executables, you can specify
the --program-prefix option to configure with a prefix ending in a
non-alphabetic character, for example "m.", and then run the programs as
m.mcron, m.cron (or m.crond) and m.crontab.
----------------------------------------------------------------------


After compilation, read the info file for full instructions for use (typing
'info -f doc/mcron.info' at the command line should suffice).  Notes for end
users, sysadmins, and developers who wish to incorporate mcron into their own
programs are included here.

Features which might be implemented sometime sooner or later are noted in the
TODO file.

Please send all other bug reports to bug-mcron@gnu.org.

Mcron is free software. See the file COPYING for copying conditions.

The mcron development home page is at http://www.gnu.org/software/mcron, and it
can be obtained from ftp://ftp.gnu.org/pub/gnu/mcron.

-----

Copyright © 2003, 2005, 2006, 2012, 2014 Dale Mellor
Copyright © 2018 Mathieu Lirzin

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.  This file is offered as-is,
without warranty of any kind.


NEWS
GNU Mcron NEWS                                    -*- outline -*-

* Noteworthy changes in release 1.2.1 (2021-08-05) [stable]

** Improvements
  Fixed "mcron --help" to show --stdin does not apply to files.
  Updated to latest autoconf.
  Changes in strings to reflect GNU Mcron and not just mcron.

* Noteworthy changes in release 1.2.0 (2020-04-22) [stable]

** Improvements
  C code removed, mcron becomes 100% Guile.
  Make doc/mcron.texi gender neutral.
  Have src/mcron/scripts/mcron.scm (process-user-file): use read and eval
    instead of load.
  New tests added for extra checks.

* Noteworthy changes in release 1.1.4 (2020-04-12) [stable]

** Improvements
  Added missing #include directives
  Support Guile 3.0
  Call 'child-cleanup' when 'select' returns an empty set
  Avoid 'call-with-current-continuation'
  Date changes for Copyrights changed for 2020
  Email updates in documentation

* Noteworthy changes in release 1.1.3 (2019-11-17) [stable]

** Improvements
  Package contains configure script by default
  Authors file change (addition)
  Doc fix for 'every second sunday'
  guix.scm update

* Noteworthy changes in release 1.1.2 (2018-11-26) [stable]

** Improvements

  The "--with-sendmail" configure variable has been added to allow the usage
  of a different Mail Transfert Agent (MTA) than 'sendmail -t'.  The MTA must
  be able to guess the actual recipients from the 'To:' message header.

* Noteworthy changes in release 1.1.1 (2018-04-08) [stable]

** Bug fixes

  The "--disable-multi-user" configure variable is not reversed anymore.
  'cron' and 'crontab' are now installed unless this option is used.

  The programs now sets the GUILE_LOAD_PATH and GUILE_LOAD_COMPILED_PATH
  environment variables with the location of the installed Guile modules.

  'next-year-from', 'next-year', 'next-month-from', 'next-month',
  'next-day-from', 'next-day', 'next-hour-from', 'next-hour',
  'next-minute-from', 'next-minute', 'next-second-from', and 'next-second' no
  longer crashes when passing an optional argument.
  [bug introduced in mcron-1.1]

** Improvements

  Some basic tests for the installed programs can be run after 'make install'
  with 'make installcheck'.

  The configuration files are now processed using a deterministic order.

  The test suite code coverage for mcron modules is now at 66.8% in term of
  number of lines (mcron-1.1 was at 23.7%).

* Noteworthy changes in release 1.1 (2018-03-19) [stable]

** New features

  The 'job' procedure has now a '#:user' keyword argument which allows
  specifying a different user that will run it.

  Additional man pages for 'cron(8)' and 'crontab(1)' are now generated using
  GNU Help2man.

** Bug fixes

  Child process created when executing a job are now properly cleaned even
  when execution fails by using 'dynamic-wind' construct.

** Improvements

  GNU Guile 2.2 is now supported.

  Some procedures are now written using functional style and include a
  docstring.  'def-macro' usages are now replaced with hygienic macros.

  Compilation is now done using a non-recursive Makefile, supports out of tree
  builds, and use silent rules by default.

  Guile object files creation don't rely on auto-compilation anymore and are
  installed in 'site-ccache' directory.

  Jobs are now internally represented using SRFI-9 records instead of vectors.

  Changelog are generated from Git logs when generating the tarball using
  Gnulib gitlog-to-changelog script.

  A test suite is now available and can be run with 'make check'.

** Changes in behavior

  The "--enable-debug" configure variable has been removed and replaced with
  MCRON_DEBUG environment variable.

  The "--disable-multi-user" configure variable is now used to not build and
  install the 'cron' and 'crontab' programs.  It has replaced the
  "--enable-no-vixie-clobber" which had similar effect.

  (mcron core) module is now deprecated and has been superseeded by
  (mcron base).

* Noteworthy changes in release 1.0.8 (2014-04-28) [stable]

  Man page is now generated with GNU Help2man before installation and
  distributed in the tarball.

* Noteworthy changes in release 1.0.7 (2012-02-04) [stable]

  Mcron is now compatible with Guile 2.0.

  FreeDesktop.org's standard user configuration directories are now used to
  find the user script files.

* Noteworthy changes in release 1.0.6 (2010-06-20) [stable]

  The copyright notices are now standardized on all auxiliary files. This
  follows the example set by the GNU hello program.

  immutable end texts from the texinfo document are now removed, to
  accomodate with Debian requirements.

* Noteworthy changes in release 1.0.5 (2010-06-13) [stable]

  Some technical changes to the build system has been made to help the
  distribution to Debian.

  The Git repository has been completely re-hashed, and now represents a
  complete and faithful history of the package's development since its
  inception.

* Noteworthy changes in release 1.0.4 (2008-02-21) [stable]

  The source code is now held in a Git repository, which can be checked-out at
  <git://git.savannah.gnu.org/mcron.git>.

  The code is now covered by the GPLv3 license.

* Noteworthy changes in release 1.0.3 (2006-04-16) [stable]

  daylight savings time shifts are now properly handled

  Parsing Vixie-style input files has been improved.

  Crontab entries can now be corrected instead of just wiping out the file.

  Mcron is now compatible with Guile 1.8.

  The manual is now licensed under the GNU Free Documentation License (GFDL)

* Noteworthy changes in release 1.0.2 (2006-01-02) [stable]

* Noteworthy changes in release 1.0.1 (2004-05-15) [stable]


main.scm
;;   Copyright (C) 2003, 2012 Dale Mellor
;; 
;;   This file is part of GNU mcron.
;;
;;   GNU mcron is free software: you can redistribute it and/or modify it under
;;   the terms of the GNU General Public License as published by the Free
;;   Software Foundation, either version 3 of the License, or (at your option)
;;   any later version.
;;
;;   GNU mcron is distributed in the hope that it will be useful, but WITHOUT
;;   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
;;   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
;;   more details.
;;
;;   You should have received a copy of the GNU General Public License along
;;   with GNU mcron.  If not, see <http://www.gnu.org/licenses/>.



;; This is the 'main' routine for the whole system; the top of this file is the
;; global entry point (after the minimal C wrapper, mcron.c.template); to all
;; intents and purposes the program is pure Guile and starts here.
;;
;; This file is built into mcron.c.template by the makefile, which stringifies
;; the whole lot, and escapes quotation marks and escape characters
;; accordingly. Bear this in mind when considering literal multi-line strings.
;;
;; (l0ad "crontab.scm") (sic) is inlined by the makefile. All other
;; functionality comes through modules in .../share/guile/site/mcron/*.scm.



;; Pull in some constants set by the builder (via autoconf) at configuration
;; time. Turn debugging on if indicated.

(use-modules (mcron config))
(if config-debug (begin (debug-enable 'debug)
                        (debug-enable 'backtrace)))



;; To determine the name of the program, scan the first item of the command line
;; backwards for the first non-alphabetic character. This allows names like
;; in.cron to be accepted as an invocation of the cron command.

(use-modules (ice-9 regex) (ice-9 rdelim))

(define command-name (match:substring (regexp-exec (make-regexp "[[:alpha:]]*$")
                                                   (car (command-line)))))



;; Code contributed by Sergey Poznyakoff.  Print an error message (made up from
;; the parts of rest), and if the error is fatal (present and non-zero) then
;; exit to the system with this code.

(define (mcron-error exit-code . rest)
  (with-output-to-port (current-error-port)
    (lambda ()
      (for-each display (append (list command-name ": ") rest))
      (newline)))
  (if (and exit-code (not (eq? exit-code 0)))
      (primitive-exit exit-code)))



;; Code contributed by Sergey Poznyakoff.  Execute body. If an 'mcron-error
;; exception occurs, print its diagnostics and exit with its error code.

(defmacro catch-mcron-error (. body)
  `(catch 'mcron-error
          (lambda ()
            ,@body)
          (lambda (key exit-code . msg)
            (apply mcron-error exit-code msg))))



;; We will be doing a lot of testing of the command name, so it makes sense to
;; perform the string comparisons once and for all here.

(define command-type (cond ((string=? command-name "mcron") 'mcron)
                           ((or (string=? command-name "cron")
                                (string=? command-name "crond")) 'cron)
                           ((string=? command-name "crontab") 'crontab)
                           (else
                            (mcron-error 12 "The command name is invalid."))))



;; There are a different set of options for the crontab personality compared to
;; all the others, with the --help and --version options common to all the
;; personalities.

(use-modules (ice-9 getopt-long))

(define options
  (catch
   'misc-error
   (lambda ()
     (getopt-long (command-line)
                  (append
                   (case command-type
                     ((crontab)
                      '((user     (single-char #\u) (value #t))
                        (edit     (single-char #\e) (value #f))
                        (list     (single-char #\l) (value #f))
                        (remove   (single-char #\r) (value #f))))
                     (else `((schedule (single-char #\s) (value #t)
                                       (predicate
                                        ,(lambda (value)
                                           (string->number value))))
                             (daemon   (single-char #\d) (value #f))
                             (noetc    (single-char #\n) (value #f))
                             (stdin    (single-char #\i) (value #t)
                                       (predicate
                                        ,(lambda (value)
                                           (or (string=? "vixie" value)
                                               (string=? "guile" value))))))))
                   '((version  (single-char #\v) (value #f))
                     (help     (single-char #\h) (value #f))))))
   (lambda (key func fmt args . rest)
     (mcron-error 1 (apply format (append (list #f fmt) args))))))

;; If the user asked for the version of this program, give it to him and get
;; out.

(if (option-ref options 'version #f)
    (begin
      (display (string-append "\n
" command-name "  (" config-package-string ")\n
Written by Dale Mellor\n
\n
Copyright (C) 2003, 2006  Dale Mellor\n
This is free software; see the source for copying conditions.  There is NO\n
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n
"))
      (quit)))



;; Likewise if the user requested the help text.

(if (option-ref options 'help #f)
    (begin
      (display (string-append "
Usage: " (car (command-line))
(case command-type

  ((mcron)
" [OPTIONS] [FILES]\n
Run an mcron process according to the specifications in the FILES (`-' for\n
standard input), or use all the files in ~/.config/cron (or the \n
deprecated ~/.cron) with .guile or .vixie extensions.\n
\n
  -v, --version             Display version\n
  -h, --help                Display this help message\n
  -sN, --schedule[=]N       Display the next N jobs that will be run by mcron\n
  -d, --daemon              Immediately detach the program from the terminal\n
                              and run as a daemon process\n
  -i, --stdin=(guile|vixie) Format of data passed as standard input or\n
                              file arguments (default guile)")

  ((cron)
" [OPTIONS]\n
Unless an option is specified, run a cron daemon as a detached process, \n
reading all the information in the users' crontabs and in /etc/crontab.\n
\n
  -v, --version             Display version\n
  -h, --help                Display this help message\n
  -sN, --schedule[=]N       Display the next N jobs that will be run by cron\n
  -n, --noetc               Do not check /etc/crontab for updates (HIGHLY\n
                              RECOMMENDED).")
  
  ((crontab)
           (string-append " [-u user] file\n"
           "       " (car (command-line)) " [-u user] { -e | -l | -r }\n"
           "               (default operation is replace, per 1003.2)\n"
           "       -e      (edit user's crontab)\n"
           "       -l      (list user's crontab)\n"
           "       -r      (delete user's crontab)\n"))

  (else "rubbish"))

"\n\n
Report bugs to " config-package-bugreport ".\n
"))
      (quit)))



;; This is called from the C front-end whenever a terminal signal is
;; received. We remove the /var/run/cron.pid file so that crontab and other
;; invocations of cron don't get the wrong idea that a daemon is currently
;; running.

(define (delete-run-file)
  (catch #t (lambda () (delete-file config-pid-file)
                       (delete-file config-socket-file))
            noop)
  (quit))



;; Setup the cron process, if appropriate. If there is already a
;; /var/run/cron.pid file, then we must assume a cron daemon is already running
;; and refuse to start another one.
;;
;; Otherwise, clear the MAILTO environment variable so that output from cron
;; jobs is sent to the various users (this may still be overridden in the
;; configuration files), and call the function in the C wrapper to set up
;; terminal signal responses to vector to the procedure above. The PID file will
;; be filled in properly later when we have forked our daemon process (but not
;; done if we are only viewing the schedules).

(if (eq? command-type 'cron)
    (begin
      (if (not (eqv? (getuid) 0))
          (mcron-error 16
                       "This program must be run by the root user (and should "
                       "have been installed as such)."))
      (if (access? config-pid-file F_OK)
          (mcron-error 1
		       "A cron daemon is already running.\n"
		       "  (If you are sure this is not true, remove the file\n"
		       "   "
		       config-pid-file
		       ".)"))
      (if (not (option-ref options 'schedule #f))
          (with-output-to-file config-pid-file noop))
      (setenv "MAILTO" #f)
      (c-set-cron-signals)))



;; Define the functions available to the configuration files. While we're here,
;; we'll get the core loaded as well.

(use-modules (mcron core)
             (mcron job-specifier)
             (mcron vixie-specification))



;; Procedure to slurp the standard input into a string.

(define (stdin->string)
  (with-output-to-string (lambda () (do ((in (read-char) (read-char)))
                                        ((eof-object? in))
                                        (display in)))))



;; Now we have the procedures in place for dealing with the contents of
;; configuration files, the crontab personality is able to validate such
;; files. If the user requested the crontab personality, we load and run the
;; code here and then get out.

(if (eq? command-type 'crontab)
    (begin
      (load "crontab.scm")
      (quit)))



;; Code contributed by Sergey Poznyakoff.  Determine if the given file is a
;; regular file or not.

(define (regular-file? file)
  (catch 'system-error
	 (lambda ()
	   (eq? (stat:type (stat file)) 'regular))
	 (lambda (key call fmt args . rest)
	   (mcron-error 0 (apply format (append (list #f fmt) args)))
	   #f)))



;; Procedure which processes any configuration file according to the
;; extension. If a file is not recognized, it is silently ignored (this deals
;; properly with most editors' backup files, for instance).

(define guile-file-regexp (make-regexp "\\.gui(le)?$"))
(define vixie-file-regexp (make-regexp "\\.vix(ie)?$"))

(define (process-user-file file-path . assume-guile)
  (cond ((string=? file-path "-")
               (if (string=? (option-ref options 'stdin "guile") "vixie")
                   (read-vixie-port (current-input-port))
                   (eval-string (stdin->string))))
        ((or (not (null? assume-guile))
             (regexp-exec guile-file-regexp file-path))
         (load file-path))
        ((regexp-exec vixie-file-regexp file-path)
         (read-vixie-file file-path))))



;; Procedure to run through all the files in a user's ~/.cron and/or
;; $XDG_CONFIG_HOME/cron or ~/.config/cron directories (only happens under the
;; mcron personality).

(define (process-files-in-user-directory)
  (let ((errors 0)
        (home-directory (passwd:dir (getpw (getuid)))))
    (map (lambda (config-directory)
          (catch #t
                 (lambda ()
                   (let ((directory (opendir config-directory)))
                     (do ((file-name (readdir directory) (readdir directory)))
                         ((eof-object? file-name) (closedir directory))
                       (process-user-file (string-append config-directory
                                                         "/"
                                                         file-name)))))
                 (lambda (key . args)
                   (set! errors (1+ errors)))))
          (list (string-append home-directory "/.cron")
                (string-append (or (getenv "XDG_CONFIG_HOME")
                                   (string-append home-directory "/.config"))
                               "/cron")))
    (if (eq? 2 errors)
        (mcron-error 13
                     "Cannot read files in your ~/.config/cron (or ~/.cron) "
                     "directory."))))



;; Procedure to check that a user name is in the passwd database (it may happen
;; that a user is removed after creating a crontab). If the user name is valid,
;; the full passwd entry for that user is returned to the caller.

(define (valid-user user-name)
  (setpwent)
  (do ((entry (getpw) (getpw)))
      ((or (not entry)
           (string=? (passwd:name entry) user-name))
       (endpwent)
       entry)))



;; Procedure to process all the files in the crontab directory, making sure that
;; each file is for a legitimate user and setting the configuration-user to that
;; user. In this way, when the job procedure is run on behalf of the
;; configuration files, the jobs are registered with the system with the
;; appropriate user. Note that only the root user should be able to perform this
;; operation, but we leave it to the permissions on the /var/cron/tabs directory
;; to enforce this.

(use-modules (srfi srfi-2))  ;; For and-let*.

(define (process-files-in-system-directory)
  (catch #t
         (lambda ()
           (let ((directory (opendir config-spool-dir)))
             (do ((file-name (readdir directory) (readdir directory)))
                 ((eof-object? file-name))
               (and-let* ((user (valid-user file-name)))
                         (set-configuration-user user)         ;; / ?? !!!!
                         (catch-mcron-error
                          (read-vixie-file (string-append config-spool-dir
                                                          "/"
                                                          file-name)))))))
         (lambda (key . args)
           (mcron-error
            4
            "You do not have permission to access the system crontabs."))))



;; Having defined all the necessary procedures for scanning various sets of
;; files, we perform the actual configuration of the program depending on the
;; personality we are running as. If it is mcron, we either scan the files
;; passed on the command line, or else all the ones in the user's .config/cron
;; (or .cron) directory. If we are running under the cron personality, we read
;; the /var/cron/tabs directory and also the /etc/crontab file.

(case command-type
  ((mcron) (if (null? (option-ref options '() '()))
                (process-files-in-user-directory)
                (for-each (lambda (file-path)
                            (process-user-file file-path #t))
                          (option-ref options '() '()))))
  
  ((cron) (process-files-in-system-directory)
   (use-system-job-list)
   (catch-mcron-error
    (read-vixie-file "/etc/crontab" parse-system-vixie-line))
   (use-user-job-list)
   (if (not (option-ref options 'noetc #f))
       (begin
         (display
"WARNING: cron will check for updates to /etc/crontab EVERY MINUTE. If you do\n
not use this file, or you are prepared to manually restart cron whenever you\n
make a change, then it is HIGHLY RECOMMENDED that you use the --noetc\n
option.\n")
         (set-configuration-user "root")
         (job '(- (next-minute-from (next-minute)) 6)
              check-system-crontab
              "/etc/crontab update checker.")))))



;; If the user has requested a schedule of jobs that will run, we provide the
;; information here and then get out.
;;
;; Start by determining the number of time points in the future that output is
;; required for. This may be provided on the command line as a parameter to the
;; --schedule option, or else we assume a default of 8. Finally, ensure that the
;; count is some positive integer.

(and-let* ((count (option-ref options 'schedule #f)))
          (set! count (string->number count))
          (display (get-schedule (if (<= count 0) 1 count)))
          (quit))
    


;; If we are supposed to run as a daemon process (either a --daemon option has
;; been explicitly used, or we are running as cron or crond), detach from the
;; terminal now. If we are running as cron, we can now write the PID file.

(if (option-ref options 'daemon (eq? command-type 'cron))
    (begin
      (if (not (eqv? (primitive-fork) 0))
          (quit))
      (setsid)
      (if (eq? command-type 'cron)
          (with-output-to-file config-pid-file
            (lambda () (display (getpid)) (newline))))))



;; If we are running as cron or crond, we establish a socket to listen for
;; updates from a crontab program. This is put into fd-list so that we can
;; inform the main wait-run-wait execution loop to listen for incoming messages
;; on this socket.

(define fd-list '())

(if (eq? command-type 'cron)
    (catch #t
           (lambda ()
             (let ((socket (socket AF_UNIX SOCK_STREAM 0)))
               (bind socket AF_UNIX config-socket-file)
               (listen socket 5)
               (set! fd-list (list socket))))
           (lambda (key . args)
             (delete-file config-pid-file)
             (mcron-error 1
                          "Cannot bind to UNIX socket "
                          config-socket-file))))
		     



;; This function is called whenever a message comes in on the above socket. We
;; read a user name from the socket, dealing with the "/etc/crontab" special
;; case, remove all the user's jobs from the job list, and then re-read the
;; user's updated file. In the special case we drop all the system jobs and
;; re-read the /etc/crontab file.

(define (process-update-request)
  (let* ((socket (car (accept (car fd-list))))
         (user-name (read-line socket)))
    (close socket)
    (set-configuration-time (current-time))
    (catch-mcron-error
     (if (string=? user-name "/etc/crontab")
         (begin
           (clear-system-jobs)
           (use-system-job-list)
           (read-vixie-file "/etc/crontab" parse-system-vixie-line)
           (use-user-job-list))
         (let ((user (getpw user-name)))
           (remove-user-jobs user)
           (set-configuration-user user)
           (read-vixie-file (string-append config-spool-dir "/" user-name)))))))



;; Added by Sergey Poznyakoff.  This no-op will collect zombie child processes
;; as soon as they die.  This is a big improvement as previously they stayed
;; around the system until the next time mcron wakes to fire a new job off.

;; Unfortunately it seems to interact badly with the select system call,
;; wreaking havoc...

;; (sigaction SIGCHLD (lambda (sig) noop) SA_RESTART)



;; Now the main loop. Forever execute the run-job-loop procedure in the mcron
;; core, and when it drops out (can only be because a message has come in on the
;; socket) we process the socket request before restarting the loop again.
;; Sergey Poznyakoff: we can also drop out of run-job-loop because of a SIGCHLD,
;; so must test fd-list.

(catch-mcron-error
 (while #t
        (run-job-loop fd-list)
        (if (not (null? fd-list))
            (process-update-request))))
